docs/clients/promtail/stages/replace.md

```diff
@@ -0,0 +1,184 @@
+map:
+
+- `time`: `2019-01-01T01:00:00.000000001Z`
+- `level`: `info`
+- `msg`: `11.11.11.11 - "POST /loki/api/push/ HTTP/1.1" 200 932 "-" "Mozilla/5.0 (Windows; U; Windows NT 5.1; de; rv:1.9.1.7) Gecko/20091221 Firefox/3.5.7 GTB6"`
+
+While the replace stage would then parse the value for `msg` in the extracted map
+and replaces the `msg` value. `msg` in extracted will now become
+
+- `msg`: `11.11.11.11 - "POST /loki/api/v1/push/ HTTP/1.1" 200 932 "-" "Mozilla/5.0 (Windows; U; Windows NT 5.1; de; rv:1.9.1.7) Gecko/20091221 Firefox/3.5.7 GTB6"`
+
+### With `replace` value in `template` format 
+
+Given the pipeline:
+
+```yaml
+- replace: 
+    expression: "^(\\S+) (\\S+) (\\S+) \\[([\\w:/]+\\s[+\\-]\\d{4})\\] \"(\\S+)\\s?(\\S+)?\\s?(\\S+)?\" (\\d{3}|-) (\\d+|-)\\s?\"?([^\"]*)\"?\\s?\"?([^\"]*)?\"?$"
+    replace: '{{ if eq .Value "200" }}{{ Replace .Value "200" "HttpStatusOk" -1 }}{{ else }}{{ .Value | ToUpper }}{{ end }}'
+```
+
+And the log line:
+
+```
+11.11.11.11 - frank [25/Jan/2000:14:00:01 -0500] "GET /1986.js HTTP/1.1" 200 932 "-" "Mozilla/5.0 (Windows; U; Windows NT 5.1; de; rv:1.9.1.7) Gecko/20091221 Firefox/3.5.7 GTB6"
+```
+
+The replace stage parses the log line and if the captured group has a value `200` it replaces the value to `HttpStatusOk`.
+
+The log line would become
+
+```
+11.11.11.11 - frank [25/Jan/2000:14:00:01 -0500] "GET /1986.js HTTP/1.1" HttpStatusOk 932 "-" "Mozilla/5.0 (Windows; U; Windows NT 5.1; de; rv:1.9.1.7) Gecko/20091221 Firefox/3.5.7 GTB6"
+```
+
+### `replace` with named captured group 
+
+Given the pipeline:
+
+```yaml
+- replace: 
+    expression: "^(?P<ip>\\S+) (?P<identd>\\S+) (?P<user>\\S+) \\[(?P<timestamp>[\\w:/]+\\s[+\\-]\\d{4})\\] \"(?P<action>\\S+)\\s?(?P<path>\\S+)?\\s?(?P<protocol>\\S+)?\" (?P<status>\\d{3}|-) (?P<size>\\d+|-)\\s?\"?(?P<referer>[^\"]*)\"?\\s?\"?(?P<useragent>[^\"]*)?\"?$"
+    replace: '{{ .Value | ToUpper }}'
+```
+
+And the log line:
+
+```
+11.11.11.11 - frank [25/Jan/2000:14:00:01 -0500] "GET /1986.js HTTP/1.1" 200 932 "-" "Mozilla/5.0 (Windows; U; Windows NT 5.1; de; rv:1.9.1.7) Gecko/20091221 Firefox/3.5.7 GTB6"
+```
+
+The replace stage parses the log line and replaces the value of all named captured groups to upper case. The
+named captured groups will be extracted to
+
+- `ip`: `11.11.11.11`
+- `identd`: `-`
+- `user`: `FRANK`
+- `timestamp`: `25/JAN/2000:14:00:01 -0500`
+- `action`: `GET`
+- `path`: `/1986.JS`
+- `protocol`: `HTTP/1.1`
+- `status`: `200`
+- `size`: `932`
+- `referer`:  `-`
+- `useragent`: `MOZILLA/5.0 (WINDOWS; U; WINDOWS NT 5.1; DE; RV:1.9.1.7) GECKO/20091221 FIREFOX/3.5.7 GTB6"`
+
+The log line would become
+
+```
+11.11.11.11 - FRANK [25/JAN/2000:14:00:01 -0500] "GET /1986.JS HTTP/1.1" 200 932 "-" "MOZILLA/5.0 (WINDOWS; U; WINDOWS NT 5.1; DE; RV:1.9.1.7) GECKO/20091221 FIREFOX/3.5.7 GTB6"
+```
+
+### `replace` with captured group and replace count
+
+Given the pipeline:
+
+```yaml
+- replace: 
+    expression: "(\\S+)"
+    replace: '$3test'
```

Comment:
Not sure what's going on here ?

Reply:
This is the syntax for providing how many matches should be replaced. 
For ex, regex: `"(\\S+)"` for a line `I am a log line` will match each of the non white space characters. So, it will match `I` `am` `a` `log` `line`. Now, if you want to replace only the first 3 matches. You can provide the syntax like this `$3test`. So, the replaced log line becomes `test test test log line`. Without the the replacement count it would be `test` and the log line would be become `test test test test test`

Reply:
I think it's fine to drop this feature. We can have multiple stage if we want to modify multiple matches. I would prefer to replace all matches and that's it at least for the first version.


In another iteration we can think about how to do the rest.

Reply:
Sure, will remove this